### PR TITLE
Support error severity screening option

### PR DESCRIPTION
**Pull request by yamarkz**
Custom Lint is a cool tool and makes my project better.
A few additional features would make the operation easier. This PR will add an option to screen Error Severity.

I implemented this issue on my own since it was a separate issue from https://github.com/invertase/dart_custom_lint/issues/101  .

Any thoughts or policies? If not, I would be happy to have a positive consideration for the merge based on this PR.
My project is going to have a constructive Custom Lint maintenance once this change is in place.

### CODE REVIEW

**Comment by yamarkz** [comment]
@rrousselGit please review and merge

**Comment by yamarkz** [comment]
@rrousselGit 

What do you think? If you do not accept the changes, please close it.

File: packages/custom_lint/lib/custom_lint.dart [hunk 1]
```diff
@@ -36,6 +37,8 @@ q: Quit
 Future<void> customLint({
   bool watchMode = true,
   required Directory workingDirectory,
+  bool fatalInfos = false,
+  bool fatalWarnings = true,
```
Comment (rrousselGit) [inline packages/custom_lint/lib/custom_lint.dart:41]: these should be required

Comment (rrousselGit) [reply packages/custom_lint/lib/custom_lint.dart:41]:
```suggestion
  required bool fatalInfos,
  required bool fatalWarnings,
```

Comment (rrousselGit) [reply packages/custom_lint/lib/custom_lint.dart:41]: Ignore this, that would be breaking. But let's make them both true

File: packages/custom_lint/lib/custom_lint.dart [hunk 2]
```diff
@@ -36,6 +37,8 @@ q: Quit
 Future<void> customLint({
   bool watchMode = true,
   required Directory workingDirectory,
+  bool fatalInfos = false,
```
Comment (rrousselGit) [inline packages/custom_lint/lib/custom_lint.dart:40]:
```suggestion
  bool fatalInfos = true,
```

File: packages/custom_lint/lib/custom_lint.dart [hunk 3]
```diff
@@ -148,20 +166,45 @@ void _renderLints(
...
-  exitCode = 1;
+  var hasErrors = false;
+  var hasWarnings = false;
+  var hasInfos = false;
   for (final error in errors) {
-    stdout.writeln(
-      '  ${_relativeFilePath(error.location.file, workingDirectory)}:${error.location.startLine}:${error.location.startColumn}'
-      ' • ${error.message} • ${error.code}',
-    );
+    hasErrors |= error.severity == AnalysisErrorSeverity.ERROR;
```
Comment (rrousselGit) [inline packages/custom_lint/lib/custom_lint.dart:182]: Don't use `|=`. This is the `|` operator, not `||`. 
So the == would be executed even though it's unnecessary.

File: packages/custom_lint/test/cli_process_test.dart [hunk 4]
```diff
@@ -72,7 +72,7 @@ void main() {
   test_app2/lib/main2.dart:1:6 • Hello world • hello_world
 ''',
     );
-    expect(process.exitCode, 1);
+    expect(process.exitCode, 0);
```
Comment (rrousselGit) [inline packages/custom_lint/test/cli_process_test.dart:75]: That's not right. If a lint if emitted, it should emit 1 not 0.

File: packages/custom_lint/test/cli_test.dart [hunk 5]
```diff
@@ -11,13 +11,22 @@ final oyPluginSource = createPluginSource([
...
     message: 'Oy',
+    errorSeverity: 'ErrorSeverity.WARNING',
   )
 ]);
 
 final helloWordPluginSource = createPluginSource([
   TestLintRule(
     code: 'hello_world',
     message: 'Hello world',
+    errorSeverity: 'ErrorSeverity.WARNING',
```
Comment (rrousselGit) [inline packages/custom_lint/test/cli_test.dart:22]: This impacts existing tests.

You should instead make a separate variable with the warning flag instead of reusing something already used by different tests but with the info status.

File: packages/custom_lint/test/cli_test.dart [hunk 6]
```diff
@@ -258,27 +334,27 @@ class _Lint extends DartLintRule {
   ) {
     final line2 = resolver.lineInfo.getOffsetOfLine(1);
     reporter.reportErrorForOffset(
-      const LintCode(name: 'x2', problemMessage: 'x2'),
+      const LintCode(name: 'x2', problemMessage: 'x2', errorSeverity: ErrorSeverity.WARNING),
```
Comment (rrousselGit) [inline packages/custom_lint/test/cli_test.dart:337]: Why these changes?

File: packages/custom_lint/test/create_project.dart [hunk 7]
```diff
@@ -37,6 +38,7 @@ class TestLintRule {
   final String onVariable;
   final String ruleMembers;
   final List<TestLintFix> fixes;
+  final String errorSeverity;
```
Comment (rrousselGit) [inline packages/custom_lint/test/create_project.dart:41]: Don't use a String for this. Either make an enum or use the official ErrorSeverity object.

**Comment by rrousselGit** [comment]
Hello! Sorry for the slow review. The notification was lost in the sea of notifications I receive :(

I had another pass at this. It's looking mostly good!  But there are some relatively minor issues. Mind fixing them?

**Comment by rrousselGit** [comment]
:wave: Do you still plan to work on this? :)

**Comment by yamarkz** [comment]
@rrousselGit 

Oops I may have misunderstood. I expected to be presented with a completely different alternative.
Can I take ownership of PR? if yes. Can I proceed with the revision according to the review comments?

**Comment by rrousselGit** [comment]
:wave:

I don't quite understand what the confusion is. Do you mind explaining?
But anyway, this PR is waiting for you to fix the various points I raised previously. Once they are fixed and I think all changes look good, this can be merged

File: packages/custom_lint/test/cli_process_test.dart (hunk 4 above)
Comment (yamarkz) [reply packages/custom_lint/test/cli_process_test.dart:68]: I'm considering the following configuration.
1. ErrorSeverity.info => Optional error (default exit code 0)
2. ErrorSeverity.warning => Default error, but optionally skippable (default exit code 1)
3. ErrorSeverity.error => Always error (default exit code 1)

This is consistent with the terminology used [in Dart analyze](https://github.com/dart-lang/sdk/blob/80bc65bb0b529b33023b4e812bd328663609921a/pkg/dartdev/lib/src/commands/analyze.dart#L44).

modifying the parameter definition [here](https://github.com/yamarkz/dart_custom_lint/blob/8fffd8d21664d4607fca402c88c9dd737d2fa476/packages/custom_lint/bin/custom_lint.dart#L12). So it will be changed to return an exit code of 0.

Could you please share your thoughts on the design for this ErrorSeverity? I would appreciate hearing your considerations and insights.

I have contemplated aligning it with the fundamental design of Dart analyzer. 
However, I am unsure if it aligns with the overall philosophy of this package.

File: packages/custom_lint/test/cli_test.dart (hunk 6 above)
Comment (yamarkz) [reply packages/custom_lint/test/cli_test.dart:337]: The reason is the same as [here](https://github.com/invertase/dart_custom_lint/pull/129#discussion_r1233210398).

File: packages/custom_lint/test/cli_test.dart (hunk 5 above)
Comment (yamarkz) [reply packages/custom_lint/test/cli_test.dart]: before considering setting up another test, I would like to confirm [this discussion](https://github.com/invertase/dart_custom_lint/pull/129#discussion_r1233210398) first.

File: packages/custom_lint/test/create_project.dart (hunk 7 above)
Comment (yamarkz) [reply packages/custom_lint/test/create_project.dart]: fixed commit 8fffd8d

File: packages/custom_lint/lib/custom_lint.dart (hunk 3 above)
Comment (yamarkz) [reply packages/custom_lint/lib/custom_lint.dart]: fixed commit 36140d7

File: packages/custom_lint/lib/custom_lint.dart [hunk 8]
```diff
@@ -148,20 +166,45 @@ void _renderLints(
...
+    hasWarnings |= error.severity == AnalysisErrorSeverity.WARNING;
+    hasInfos |= error.severity == AnalysisErrorSeverity.INFO;
+  }
+
+  if (hasErrors) {
+    exitCode = 1;
+    return;
```
[the remark this answers is not in the document]
Comment (yamarkz) [reply packages/custom_lint/lib/custom_lint.dart:187]: fixed commit 36140d7.

I didn't take enough considerations for code efficiency.

File: packages/custom_lint/test/cli_process_test.dart (hunk 4 above)
Comment (rrousselGit) [reply packages/custom_lint/test/cli_process_test.dart:68]: I would prefer basing custom_lint on `flutter analyze` rather than `dart analyze`.

With `flutter analyze`, infos lead to an exitCode 1 too, and you do `flutter analyze --no-fatal-infos`

File: packages/custom_lint/test/cli_test.dart [hunk 9]
```diff
@@ -11,13 +12,22 @@ final oyPluginSource = createPluginSource([
...
     message: 'Oy',
+    errorSeverity: ErrorSeverity.WARNING,
   )
 ]);
 
 final helloWordPluginSource = createPluginSource([
   TestLintRule(
     code: 'hello_world',
     message: 'Hello world',
+    errorSeverity: ErrorSeverity.WARNING,
```
Comment (rrousselGit) [inline packages/custom_lint/test/cli_test.dart:23]: This modifies existing tests, which is undesired.

Add a new source instead

File: packages/custom_lint/test/cli_test.dart [hunk 10]
```diff
@@ -11,13 +12,22 @@ final oyPluginSource = createPluginSource([
...
+  )
+]);
+
+final helloDartPluginSource = createPluginSource([
```
Comment (rrousselGit) [inline packages/custom_lint/test/cli_test.dart:27]: What is the use for this one?

File: packages/custom_lint/test/cli_test.dart [hunk 11]
```diff
@@ -247,7 +324,7 @@ class _HelloWorldLint extends PluginBase {
 }
 
 class _Lint extends DartLintRule {
-  const _Lint() : super(code: const LintCode(name: 'a', problemMessage: 'a'));
+  const _Lint() : super(code: const LintCode(name: 'a', problemMessage: 'a', errorSeverity: ErrorSeverity.WARNING));
```
Comment (rrousselGit) [inline packages/custom_lint/test/cli_test.dart:327]: Please format to use trailing commas

**Comment by shilangyu** [comment]
What is the status here? Is there some help needed finishing this PR? A `--no-fatal-infos` option would be very useful.

**Comment by yamarkz** [comment]
@shilangyu 
sorry, my hand was stuck. I'll resume development soon

File: packages/custom_lint/test/cli_test.dart [hunk 12]
```diff
@@ -86,6 +87,61 @@ lib/custom_lint_client.dart:13:29: Error: Undefined name 'createPlugin'.
...
+      plugins: {'test_lint': plugin.uri},
+      name: 'test_app',
+    );
+
+    await runWithIOOverride(
+      (out, err) async {
+        await cli.entrypoint(['--no-fatal-infos']);
+
+        expect(err, emitsDone);
+        expect(exitCode, 0);
```
Comment (rrousselGit) [inline packages/custom_lint/test/cli_test.dart:106]: Could you test that `out` still logs the infos/warnings?

Comment (yamarkz) [reply packages/custom_lint/test/cli_test.dart:103]: fixed commit 882b9bb

**Comment by rrousselGit** [comment]
Great thanks! I'll wait for the CI to complete and merge this

File: packages/custom_lint/test/goldens/server_test/redirect_logs.golden [hunk 13]
```diff
@@ -1,4 +1,4 @@
 [hello_world] 1990-01-01T00:00:00.000 Hello world
 [hello_world] 1990-01-01T00:00:00.000 Plugin hello_world threw while analyzing app/lib/another.dart:
 [hello_world] 1990-01-01T00:00:00.000 Bad state: fail
-[hello_world] 1990-01-01T00:00:00.000 #0      hello_world.run.<anonymous closure> (package:test_lint/test_lint.dart:26:3)
\ No newline at end of file
+[hello_world] 1990-01-01T00:00:00.000 #0      hello_world.run.<anonymous closure> (package:test_lint/test_lint.dart:28:3)
```
Comment (rrousselGit) [inline packages/custom_lint/test/goldens/server_test/redirect_logs.golden:4]:
```suggestion
[hello_world] 1990-01-01T00:00:00.000 #0      hello_world.run.<anonymous closure> (package:test_lint/test_lint.dart:28:3)```
